prov/efa/test/efa_unit_test_utils.h

```diff
@@ -0,0 +1,113 @@
+#include <cmocka.h>
+#include <infiniband/verbs.h>
+#include "efa.h"
+
+/**
+ * @brief Initialize a efa_cq struct
+ * @param[in] efa_cq A pointer to `struct efa_cq`.
+ * @returns Error code. 0 if successful, non-zero if failed.
+ */
+static int efa_cq_initialize(struct efa_cq *efa_cq)
+{
+    static struct fid_cq fid_cq = {0};
+    static struct util_cq util_cq = {0};
+    static struct efa_domain domain = {0};
+    static struct efa_fabric fabric = {0};
+    static struct util_fabric util_fabric = {0};
+    static struct fid_fabric fid_fabric = {0};
+    static struct ibv_cq_ex *ibv_cq_ex;
+
+    ibv_cq_ex = calloc(1, sizeof(struct ibv_cq_ex));
+
+    if (!ibv_cq_ex)
+        return 1;
+
+    fid_fabric.api_version = 0;
+    util_fabric.fabric_fid = fid_fabric;
+    fabric.util_fabric = util_fabric;
+    domain.fabric = &fabric;
+    efa_cq->util_cq = util_cq;
+    efa_cq->util_cq.cq_fid = fid_cq;
+    efa_cq->ibv_cq_ex = ibv_cq_ex;
+    efa_cq->domain = &domain;
+    ofi_spin_init(&efa_cq->lock);
+
+    return 0;
+}
+
+/** 
+ * @brief Teardown a struct efa_cq object initialized with efa_cq_initialize
+ * @param[in] efa_cq Pointer to a struct efa_cq object initialized with efa_cq_initialize 
+ */
+static void teardown_efa_cq(struct efa_cq *efa_cq)
+{
+    ofi_spin_destroy(&efa_cq->lock);
+    free(efa_cq->ibv_cq_ex);
+}
+
+/**
+ * @brief Initialize a efa_ep struct
+ * @param[in] efa_ep A pointer to `struct efa_ep`.
+ * @returns Error code. 0 if successful, non-zero if failed.
+ */
+static int efa_ep_initialize(struct efa_ep *efa_ep)
+{
+    static struct util_ep util_ep = {0};
+    static struct fid_ep fid_ep = {0};
+
+    util_ep.ep_fid = fid_ep;
+    efa_ep->util_ep = util_ep;
+    efa_ep->av = NULL;
+
+    return 0;
+}
+
+/**
+ * @brief Initialize a rxr_ep struct
+ * @param[in] rxr_ep A pointer to `struct rxr_ep`.
+ * @param[in] efa_cq A pointer to `struct efa_cq`. This is used to inject values into rxr_ep.
+ * @param[in] efa_ep A pointer to `struct efa_ep`. This is used to inject values into rxr_ep.
+ * @returns Error code. 0 if successful, non-zero if failed.
+ */
+static int rxr_ep_initialize(struct rxr_ep *rxr_ep, struct efa_cq *efa_cq, struct efa_ep *efa_ep)
```

Comment:
why not just call `fi_endpoint` to create an ep?

Like we did in `https://github.com/ofiwg/libfabric/blob/62521e7eef653367e3a6b88c99be7173c4e7befa/prov/efa/test/efa_unit_test_ep.c#L53`

Reply:
IIUC this creates a dependency on efa device. For unit test we can deep mock the data structures directly. With this deep mock we can run the test anywhere.